sky/cloud_stores.py

```diff
@@ -153,6 +154,135 @@ def make_sync_file_command(self, source: str, destination: str) -> str:
         return ' && '.join(all_commands)
 
 
+class AzureCloudStorage(CloudStorage):
+    """Azure Blob Storage."""
+    # AzCopy is utilized for downloading data from Azure Blob Storage
+    # containers to local systems due to its superior performance compared to
+    # az-cli. While az-cli's `az storage blob sync` can synchronize data from
+    # local to container, it lacks support for container to local
+    # synchronization. Moreover, `az storage blob download-batch` in az-cli
+    # does not leverage AzCopy's efficient multi-threaded capabilities, leading
+    # to slower performance.
+    #
+    # AzCopy requires appending SAS tokens directly in commands, as it does not
+    # support using STORAGE_ACCOUNT_KEY, unlike az-cli, which can generate
+    # SAS tokens but lacks direct multi-threading support like AzCopy.
+    # Hence, both tools are necessary: az-cli for SAS token generation and
+    # AzCopy for efficient data transfer from containers to local systems.
+    # Note that on Azure instances, both az-cli and AzCopy are typically
+    # pre-installed. And installing both would be used with AZ container is
+    # used from non-Azure instances.
+
+    # Installation reference: https://learn.microsoft.com/en-us/cli/azure/install-azure-cli-linux?pivots=apt#option-2-step-by-step-installation-instructions # pylint: disable=line-too-long
+    _AZ_CLI_VERSION = '2.58.0'
+    _GET_AZCLI = [
+        'az --version >/dev/null 2>&1 || '
+        '(sudo apt-get update; '
+        'sudo apt-get install ca-certificates curl apt-transport-https '
+        'lsb-release gnupg -y; '
+        'sudo mkdir -p /etc/apt/keyrings; '
+        'curl -sLS https://packages.microsoft.com/keys/microsoft.asc | '
+        'gpg --dearmor | '
+        'sudo tee /etc/apt/keyrings/microsoft.gpg > /dev/null; '
+        'sudo chmod go+r /etc/apt/keyrings/microsoft.gpg; '
+        'AZ_DIST=$(lsb_release -cs); '
+        'echo "deb [arch=`dpkg --print-architecture` '
+        'signed-by=/etc/apt/keyrings/microsoft.gpg] '
+        'https://packages.microsoft.com/repos/azure-cli/ $AZ_DIST main" | '
+        'sudo tee /etc/apt/sources.list.d/azure-cli.list; '
+        'sudo apt-get update; '
+        f'AZ_VER={_AZ_CLI_VERSION}; '
+        'sudo apt-get install azure-cli=$AZ_VER-1~$AZ_DIST -y)'
+    ]
```

Comment:
Thanks for pointing this out. This was originally written so that `az-cli` can be installed at the remote instance as well. Having this set up as azure dependency from `setup.py` doesn't install at the remote instance(aws or gcp), so this was necessary. I took a second look, and it turns out installing `az-cli` at the remote instance isn't necessary to run the functions. Removed the logic to install `az-cli`.

Reply:
reopened for `_GET_AZCOPY` as well. I was curious why S3 and GCS support does not need this, or if I missed sth pls point it out.

Reply:
S3 and GCS support in `cloud_stores.py` does all have similar run time installation at remote machine. 

These are necessary in case when either `aws cli`, `gsutil`, or `azcopy` are not installed in the remote machine. And this happens, for example, when the user is using different cloud providers of the remote machine and the cloud storage. Say, if the user is using AWS instance but using azure cloud storage to fetch data, then, it's necessary to install `azcopy` in runtime at the remote AWS instance.